### PR TITLE
fix: performance improvements for js editor

### LINKED ISSUE #21467: [Bug]: The app begins to slow down on the cloud version when the app becomes complex
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

when the user starts coding, performance is quite good but with time (as things become more complex), the response of their typing gets slower and slower. They have always had to wait a while until the text or code, that they write is the (https://www.loom.com/share/3cacb2cb93ec40cbb63fdc3095569246).
The app slows down everywhere: JS editor and widget editor
it was really good in the beginning and while coding their stuff it became slower and slower.

### Steps To Reproduce

1. Create a new app on the cloud
2. Build until things get very complex

### Public Sample App

_No response_

### Environment

Production

### Issue video log

_No response_

### Version

Cloud 1.9.9

Front logo [Front conversations](https://app.frontapp.com/open/top_85wp8)

### PR DESCRIPTION
**Pull request by eco-monk**
## Description

TL;DR performance improvements for js editor

- fix entityNavigationData generation (to prevent unnecessary component updates)
  - in codeEditor/index.ts (`addThisReference` was creating a new object everytime)
  - in navigationSelector.ts (use `getJSCollections` instead of `getJSCollectionsForCurrentPage`, which created a new object everytime, even if actions were not updated)
- combine markers for navigation and peek overlay to reduce the total number of markers
- clear and add marks for only the edited lines instead of the whole file

Note: once a js object is saved, it's still going to trigger a whole file clear and marking.
Because, it's an entity update which needs a whole refresh of the markers.

Fixes #21467

## Media
Case: Adding a blank space in js editor.

#### Before: 
image

####After:
image

---

### Reduced entity marker called count:
https://www.loom.com/share/23719f8dfde8457ea0a86f44500ec34a

---

#### Before:
image

#### After:
image

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Manual

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/components/editorComponents/CodeEditor/MarkHelpers/bindingMarker.ts
Comment (eco-monk) [inline app/client/src/components/editorComponents/CodeEditor/MarkHelpers/bindingMarker.ts:1]: Existing file split up,
No changes here.

File: app/client/src/components/editorComponents/CodeEditor/MarkHelpers/entityMarker.ts
Comment (eco-monk) [inline app/client/src/components/editorComponents/CodeEditor/MarkHelpers/entityMarker.ts:1]: existing file split + changes.
Change commit [here](https://github.com/appsmithorg/appsmith/pull/21492/commits/71bbb17adb6b56d1b50f3a58b3e8b438de5047e2#diff-ff3b30c0a2202f90f94ff7a6ebccdd090cf40097b65d7b67098e4377cf1df619).

**Review by eco-monk — COMMENTED** [review]
Summary

File: app/client/src/selectors/navigationSelectors.ts [hunk 1]
```diff
@@ -36,11 +36,22 @@ export type EntityNavigationData = Record<string, NavigationData>;
...
-  (actions, plugins, jsActions, widgets, pageId, dataTree: DataTree) => {
+  (_: any, jsObjectName: string | undefined) => jsObjectName,
+  (
+    actions,
+    plugins,
+    jsActions,
+    widgets,
+    pageId,
+    dataTree: DataTree,
+    jsObjectName: string | undefined,
```
Comment (hetunandu) [inline app/client/src/selectors/navigationSelectors.ts:51]: Except entity name here. You can choose to only check this in the JS Object loop

File: app/client/src/selectors/navigationSelectors.ts [hunk 2]
```diff
@@ -36,11 +36,22 @@ export type EntityNavigationData = Record<string, NavigationData>;
 export const getEntitiesForNavigation = createSelector(
   getActionsForCurrentPage,
   getPlugins,
-  getJSCollectionsForCurrentPage,
+  getJSCollections, // don't use getJSCollectionsForCurrentPage (returns a new object everytime)
```
Comment (hetunandu) [inline app/client/src/selectors/navigationSelectors.ts:39]: Comment should be part of the `getJSCollectionsForCurrentPage` selector

File: app/client/src/components/editorComponents/CodeEditor/index.tsx [hunk 3]
```diff
@@ -1361,12 +1370,10 @@ const mapStateToProps = (state: AppState, props: EditorProps) => ({
...
   ),
-  entitiesForNavigation: props.isJSObject
-    ? addThisReference(
-        getEntitiesForNavigation(state),
-        props.dataTreePath?.split(".")[0],
-      )
-    : getEntitiesForNavigation(state),
+  entitiesForNavigation: getEntitiesForNavigation(
+    state,
+    props.isJSObject ? props.dataTreePath?.split(".")[0] : undefined,
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/CodeEditor/index.tsx:1375]: pass the entity name every time

**Review by hetunandu — CHANGES_REQUESTED** [review]
Minor change request

**Comment by eco-monk** [comment]
Pending QA verification.

**Comment by Richarex** [comment]
Tested this on the DP url, working as expected.
